### PR TITLE
Torch frontend lerp

### LINKED ISSUE #10683: lerp

### PR DESCRIPTION
**Pull request by aparna2071**
Closes #10683

### CODE REVIEW

File: ivy/functional/frontends/torch/pointwise_ops.py [hunk 1]
```diff
@@ -450,3 +450,8 @@ def hypot(input, other, *, out=None):
 @to_ivy_arrays_and_back
 def sigmoid(input, *, out=None):
     return ivy.sigmoid(input, out=out)
+
+
+@to_ivy_arrays_and_back
+def lerp(start, end, weight):
+    return ivy.lerp(start, end, weight)
```
Comment (fspyridakos) [inline ivy/functional/frontends/torch/pointwise_ops.py:457]: ivy.lerp doesn't exist in the ivy API, you can suggest its addition by following the steps here: https://lets-unify.ai/ivy/deep_dive/ivy_frontends.html#missing-ivy-functions.
That being said the function can simply composed by a few arithmetic operations as described here: https://pytorch.org/docs/stable/generated/torch.lerp.html.
Also please add the `out` argument

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 2]
```diff
@@ -6,6 +6,9 @@
 import ivy_tests.test_ivy.helpers as helpers
 from ivy_tests.test_ivy.helpers import handle_frontend_test
 
+import torch
+from torch.testing import assert_allclose
+
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:11]: Please remove these imports

File: ivy/functional/frontends/torch/pointwise_ops.py [hunk 3]
```diff
@@ -450,3 +450,8 @@ def hypot(input, other, *, out=None):
 @to_ivy_arrays_and_back
 def sigmoid(input, *, out=None):
     return ivy.sigmoid(input, out=out)
+
+
+@to_ivy_arrays_and_back
+def lerp(start, end, weight):
```
Comment (fspyridakos) [inline ivy/functional/frontends/torch/pointwise_ops.py:456]: The first argument should be `input` as stated in the torch docs: https://pytorch.org/docs/stable/generated/torch.lerp.html

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 4]
```diff
@@ -2081,3 +2084,37 @@ def test_torch_sigmoid(
...
+    fn_tree,
+    on_device,
+):
+    input_dtype, inputs = dtype_and_input
+    start, end, weight = inputs
+
+    expected_output = torch.lerp(start, end, weight)
+    output = frontend.torch.lerp(start, end, weight)
+
+    assert_allclose(expected_output, output, rtol=1e-03, atol=1e-05)
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2111]: No need to do this here, `test_frontend_fuction` handles this

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 5]
```diff
@@ -2081,3 +2084,37 @@ def test_torch_sigmoid(
...
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        input=(start, end, weight),
+        expected_output=expected_output,
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2119]: don't pack all input arguments into one, make them seperate keyword args, i.e. `input=start`, `end=end`, `weight=weight`.
Remove `expected_output`
No need to include the `out` argument when you add it

**Review by fspyridakos — CHANGES_REQUESTED** [review]
Thanks for your contribution! I left a few comments please go over them and ask for another review once you're done.

**Comment by aparna2071** [comment]
Thank you so much for your detailed review, will help me a lot.
I will make the necessary changes as soon as possible.

**Comment by aparna2071** [comment]
Hi @fspyridakos, as far as the addition of ivy.lerp function is concerned, I've raised a new issue #10762.
Other than that, I've tried to make all necessary changes as suggested by you.
Kindly review & please feel free to suggest if any more changes are required.
Also, thanks a lot for detailed suggestions.

File: ivy/functional/frontends/torch/pointwise_ops.py [hunk 6]
```diff
@@ -450,3 +450,11 @@ def hypot(input, other, *, out=None):
 @to_ivy_arrays_and_back
 def sigmoid(input, *, out=None):
     return ivy.sigmoid(input, out=out)
+
+
+@to_ivy_arrays_and_back
+def lerp(input, end, weight, *, out=None):
+    if out is None:
+        out = ivy.zeros_like(input)
+    out = ivy.multiply_add(ivy.subtract(end, input), weight, ivy.subtract(input, out), out=out)
```
Comment (fspyridakos) [inline ivy/functional/frontends/torch/pointwise_ops.py:459]: most lines are redundant, you can just do `return <operation>` as with the above functions (`out` in the docs practically means the `return`) also look over your implementation, `ivy.multiply_add` doesn't exist

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 7]
```diff
@@ -2081,3 +2081,33 @@ def test_torch_sigmoid(
...
         input=x[0],
     )
+
+
+# lerp
+@handle_frontend_test(
+    fn_tree="torch.lerp",
+    dtype_and_input=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        default=0.5,
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2091]: I'm not sure what `default` is here, but you should probably use `num_arrays=3`

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 8]
```diff
@@ -2081,3 +2081,33 @@ def test_torch_sigmoid(
...
+helpers.test_frontend_function(
+input_dtypes=input_dtype,
+frontend=frontend,
+test_flags=test_flags,
+fn_tree=fn_tree,
+on_device=on_device,
+input_start=start,
+input_end=end,
+input_weight=weight,
+)
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2113]: fix indentation here to match functions above

**Review by fspyridakos — CHANGES_REQUESTED** [review]
Thanks for the changes! Also your ivy.lerp issue mentions the `.interp` function in jax and numpy, and `ivy.interp` already exists, if you think you can use that function here go ahead, but I think `lerp` and `interp` do different things, just chaning some elementwise functions is fine for now

**Comment by aparna2071** [comment]
Hi @fspyridakos, once again thanks a lot.
As far as ivy.lerp issue is concerned, I've deleted the links for interp & have tried to make all necessary changes as required for this PR.
Kindly review & please let me know if any more changes are required.

File: ivy/functional/frontends/torch/pointwise_ops.py [hunk 9]
```diff
@@ -450,3 +450,9 @@ def hypot(input, other, *, out=None):
 @to_ivy_arrays_and_back
 def sigmoid(input, *, out=None):
     return ivy.sigmoid(input, out=out)
+
+
+@to_ivy_arrays_and_back
```
Comment (fspyridakos) [inline ivy/functional/frontends/torch/pointwise_ops.py:455]: Please add float16 and bfloat16 as unsupprted dtypes for torch (like other functions above)

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 10]
```diff
@@ -2081,3 +2081,33 @@ def test_torch_sigmoid(
...
         input=x[0],
     )
+
+
+# lerp
+@handle_frontend_test(
+    fn_tree="torch.lerp",
+    dtype_and_input=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        num_arrays=3,
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2091]: Add the following:
`       large_abs_safety_factor=2.5,
        small_abs_safety_factor=2.5,
        safety_factor_scale="log",`
Some large values will not interpolate properly due to float percision limitations, this prevents those from being generated

File: ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py [hunk 11]
```diff
@@ -2081,3 +2081,33 @@ def test_torch_sigmoid(
...
+    start, end, weight = inputs
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        input_start=start,
+        input_end=end,
+        input_weight=weight,
```
Comment (fspyridakos) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_pointwise_ops.py:2112]: just have this be `input=start, end=end` and so on, the kwargs must have the same names as the function args

**Review by fspyridakos — CHANGES_REQUESTED** [review]
Few more chnages and you're golden

**Comment by aparna2071** [comment]
Hi @fspyridakos, thanks for all your help.
I've made the changes, please let me know if anything else is required.

**Comment by fspyridakos** [comment]
LGTM, had to make 2 small changes but it's good now and the test passes

**Comment by aparna2071** [comment]
Thanks a lot for merging @fspyridakos.

**Pull request merged**